### PR TITLE
Implement progress endpoints & display them in components

### LINKED ISSUE #39: Chapter progress
- [x] CMS Block
- [x] LMS Endpoint to get course chapter progress
- [x] Material-template component

Pending: Planning & Implementation of unlocking exercises/parts based on something

### LINKED ISSUE #425: Fetch whole course progress for student endpoint & Implement fetching points in shared-module/CourseProgress component

### LINKED ISSUE #426: Fetch user exercise progress with exercise_id endpoint & Implement fetching in ExercisesInThisChapter component

### PR DESCRIPTION
**Pull request by sebazai**
Resolves #426 
Resolves #425 
Resolves #39

### CODE REVIEW

File: services/headless-lms/src/models/chapters.rs [hunk 1]
```diff
@@ -345,3 +353,26 @@ RETURNING *;
...
+    chapter_id: &Uuid,
+    user_id: &Uuid,
+) -> ModelResult<UserChapterProgress> {
+    let mut connection = pool.acquire().await?;
+
+    let mut exercises =
+        crate::models::exercises::get_exercises_by_chapter_id(&mut connection, chapter_id).await?;
+
+    let exercise_ids: Vec<Uuid> = exercises.iter_mut().map(|e| e.id).collect();
+    let exercise_points: Vec<i32> = exercises.into_iter().map(|e| e.score_maximum).collect();
```
Comment (nygrenh) [inline services/headless-lms/src/models/chapters.rs:368]:
```suggestion
    let score_maximum: i32 = exercises.into_iter().map(|e| e.score_maximum).sum();
```

File: services/headless-lms/src/models/user_exercise_states.rs [hunk 2]
```diff
@@ -95,17 +104,43 @@ WHERE ues.course_instance_id = $1
...
+pub async fn get_user_chapter_metrics(
+    pool: &PgPool,
+    exercise_ids: &[Uuid],
+    user_id: &Uuid,
+) -> ModelResult<UserChapterMetrics> {
+    let mut connection = pool.acquire().await?;
+    let res = sqlx::query_as!(
+        UserChapterMetrics,
+        r#"
+SELECT COALESCE(SUM(ues.score_given), 0) AS score_given
```
Comment (nygrenh) [inline services/headless-lms/src/models/user_exercise_states.rs:116]: Only from a specific course instance

File: services/headless-lms/src/models/user_exercise_states.rs [hunk 3]
```diff
@@ -95,17 +104,43 @@ WHERE ues.course_instance_id = $1
...
+FROM user_exercise_states AS ues
+WHERE ues.exercise_id IN (
+    SELECT UNNEST($1::uuid [])
+  )
+  AND ues.deleted_at IS NULL
+  AND ues.user_id = $2
+  AND ues.activity_progress IN ('submitted', 'completed');
```
Comment (nygrenh) [inline services/headless-lms/src/models/user_exercise_states.rs:123]:
```suggestion
```

File: services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/PageExerciseList.tsx [hunk 4]
```diff
@@ -5,9 +5,14 @@ import PageBox from "../../../shared-module/components/ExerciseList/PageBox"
...
+  courseInstanceId: string
 }
 
-const PageExerciseList: React.FC<PageExerciseListProps> = ({ page, courseSlug }) => {
+const PageExerciseList: React.FC<PageExerciseListProps> = ({
+  page,
+  courseSlug,
+  courseInstanceId,
+}) => {
   return (
```
Comment (nygrenh) [inline services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/PageExerciseList.tsx:16]: useQuery(`chapter-exercise-list-${page.chapter_id}-${course_instance_id}`

endpoint return map { exercise_id: score info }

File: services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/PageExerciseList.tsx [hunk 5]
```diff
@@ -5,9 +5,14 @@ import PageBox from "../../../shared-module/components/ExerciseList/PageBox"
 export interface PageExerciseListProps {
   page: PageWithExercises
   courseSlug: string
+  courseInstanceId: string
 }
 
-const PageExerciseList: React.FC<PageExerciseListProps> = ({ page, courseSlug }) => {
+const PageExerciseList: React.FC<PageExerciseListProps> = ({
```
Comment (nygrenh) [inline services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/PageExerciseList.tsx:11]: Rename to ChapterExerciseListGroupedByPage

File: services/headless-lms/src/controllers/course_material/course_instances.rs [hunk 6]
```diff
@@ -72,39 +72,38 @@ async fn get_user_progress_for_course_instance_chapter(
...
 Response:
 ```json
 {
+    "exercise_id": "uuid"
     "score_given":1.0
 }
  ```
 */
-async fn get_user_progress_for_course_instance_exercise(
+async fn get_user_progress_for_course_instance_chapter_exercise(
```
Comment (nygrenh) [inline services/headless-lms/src/controllers/course_material/course_instances.rs]: exercises

File: services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/ChapterExerciseListGroupedByPage.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        <pre>{JSON.stringify(error, undefined, 2)}</pre>
+      </div>
+    )
+  }
+
+  if (isLoading || !data) {
+    return <div>Loading exercise user progress...</div>
+  }
+
+  const mappedExerciseUserScore = new Map(data.map((x) => [x.exercise_id, x.score_given]))
```
Comment (jolampi) [inline services/course-material/src/components/ContentRenderer/ExerciseInChapterBlock/ChapterExerciseListGroupedByPage.tsx:39]: Map data in `fetchUserCourseInstanceExerciseProgress` before returning it?

**Pull request merged**